File: pipeline/subtitle-gen.ts
#!/usr/bin/env bun
/**
 * Generate Thai SRT subtitle from Whisper timing + translation text
 *
 * Flow:
 * 1. Convert TTS audio to 16kHz WAV
 * 2. Run mlx_whisper turbo → get segments with timestamps
 * 3. SequenceMatcher align whisper text → translation text
 * 4. Output SRT with correct timing + correct text (no periods)
 *
 * Usage: bun pipeline/subtitle-gen.ts <date> <mosque>
 */

import { execSync } from "child_process";
import { existsSync, readFileSync, writeFileSync, mkdirSync } from "fs";
import { parse as parseYaml } from "yaml";
import { join } from "path";

const ROOT = import.meta.dir.replace("/pipeline", "");
const CONFIG = parseYaml(readFileSync(join(ROOT, "config.yaml"), "utf-8"));

/** Format seconds to SRT timestamp */
function formatSrtTime(seconds: number): string {
  const h = Math.floor(seconds / 3600);
  const m = Math.floor((seconds % 3600) / 60);
  const s = Math.floor(seconds % 60);
  const ms = Math.floor((seconds % 1) * 1000);
  return `${String(h).padStart(2, "0")}:${String(m).padStart(2, "0")}:${String(s).padStart(2, "0")},${String(ms).padStart(3, "0")}`;
}

/** Run Whisper on TTS audio to get timed segments */
function runWhisper(audioPath: string, outputDir: string): any {
  // Convert to 16kHz WAV for speed
  const wavPath = join(outputDir, "whisper_input.wav");
  if (!existsSync(wavPath)) {
    execSync(`ffmpeg -y -i "${audioPath}" -ar 16000 -ac 1 "${wavPath}" 2>/dev/null`, { timeout: 30000 });
  }

  // Run mlx_whisper (Apple Silicon optimized)
  const jsonName = "whisper_input.json";
  const jsonPath = join(outputDir, jsonName);
  if (!existsSync(jsonPath)) {
    console.log("     🎤 Running mlx_whisper turbo...");
    execSync(
      `mlx_whisper "${wavPath}" --model mlx-community/whisper-turbo --language th -f json -o "${outputDir}"`,
      { timeout: 300000 }
    );
  }

  return JSON.parse(readFileSync(jsonPath, "utf-8"));
}

/** Get clean translation text (same strip as TTS but remove periods) */
function getCleanText(date: string, mosque: string): string {
  const translationPath = join(ROOT, "content", date, mosque, "translation-th.md");
  let text = readFileSync(translationPath, "utf-8");

  // Remove frontmatter
  if (text.startsWith("---\n")) {
    const endIdx = text.indexOf("\n---\n", 4);
    if (endIdx > 0) text = text.slice(endIdx + 5);
  }

  text = text
    .replace(/^#{1,6}\s+.*$/gm, "")
    .replace(/^---+$/gm, "")
    .replace(/.*ส่วนที่\s+\d+\/\d+.*/g, "")
    .replace(/^-?\s*\*{0,2}(วันที่|เชค|แหล่งที่มา|Date|Sheikh|Source).*$/gim, "")
    .replace(/https?:\/\/\S+/g, "")
    .replace(/\[([^\]]*)\]\([^)]*\)/g, "$1")
    .replace(/## ไฮไลท์สำหรับ Reels[\s\S]*$/m, "")
    .replace(/\*\*(.+?)\*\*/g, "$1")
    .replace(/\*(.+?)\*/g, "$1")
    .replace(/^>\s*/gm, "")
    .replace(/^[-*]\s+/gm, "")
    // Remove periods (for display, not TTS)
    .replace(/\.(\n)/g, "$1")
    .replace(/\. /g, " ")
    .replace(/\.$/gm, "")
    .replace(/\n{3,}/g, "\n\n")
    .replace(/\s{2,}/g, " ")
    .trim();

  // Flatten to single line for alignment
  return text.replace(/\n+/g, " ").trim();
}

/** Sequence alignment — map whisper character positions to script positions */
function alignTexts(whisperText: string, scriptText: string): Map<number, number> {
  const w2s = new Map<number, number>();

  // Simple LCS-based alignment (character level)
  const wLen = whisperText.length;
  const sLen = scriptText.length;

  // Use sliding window matching for efficiency
  let wi = 0;
  let si = 0;

  while (wi < wLen && si < sLen) {
    if (whisperText[wi] === scriptText[si]) {
      w2s.set(wi, si);
      wi++;
      si++;
    } else {
      // Try to find match nearby
      let foundW = -1, foundS = -1;
      const searchRange = 20;

      // Look ahead in script
      for (let j = 1; j <= searchRange && si + j < sLen; j++) {
        if (whisperText[wi] === scriptText[si + j]) {
          foundS = si + j;
          break;
        }
      }
      // Look ahead in whisper
      for (let j = 1; j <= searchRange && wi + j < wLen; j++) {
        if (whisperText[wi + j] === scriptText[si]) {
          foundW = wi + j;
          break;
        }
      }

      if (foundS >= 0 && (foundW < 0 || foundS - si <= foundW - wi)) {
        // Skip in script (insertion)
        for (let j = si; j < foundS; j++) {
          w2s.set(wi, j);
        }
        si = foundS;
      } else if (foundW >= 0) {
        // Skip in whisper (deletion)
        for (let j = wi; j < foundW; j++) {
          w2s.set(j, si);
        }
        wi = foundW;
      } else {
        // No match — map proportionally
        w2s.set(wi, si);
        wi++;
        si++;
      }
    }
  }

  return w2s;
}

/** Main subtitle generation */
export function generateSubtitle(date: string, mosque: string): string | null {
  const audioDir = join(ROOT, CONFIG.output.audio_dir);
  const ttsAudio = join(audioDir, `${date}-${mosque}-th.mp3`);

  if (!existsSync(ttsAudio)) {
    console.log("  ❌ TTS audio not found");
    return null;
  }

  // Step 1: Run Whisper
  const whisperDir = join(audioDir, `${date}-${mosque}-whisper`);
  mkdirSync(whisperDir, { recursive: true });
  const whisperResult = runWhisper(ttsAudio, whisperDir);
  const segments = whisperResult.segments;
  console.log(`     📊 Whisper: ${segments.length} segments`);

  // Step 2: Get clean translation text
  const scriptText = getCleanText(date, mosque);
  console.log(`     📝 Script: ${scriptText.length} chars`);

  // Step 3: Align whisper text → script text
  console.log("     🔗 Aligning whisper ↔ script...");
  const whisperFullText = segments.map((s: any) => s.text.trim()).join("");
  const w2s = alignTexts(whisperFullText, scriptText);

  // Step 4: Map segments to corrected text
  const corrected: { text: string; start: number; end: number }[] = [];
  let wp = 0;

  for (const seg of segments) {
    const segText = seg.text.trim();
    const segLen = segText.length;
    const mappedPositions: number[] = [];

    for (let i = wp; i < wp + segLen; i++) {
      const mapped = w2s.get(i);
      if (mapped !== undefined) mappedPositions.push(mapped);
    }

    if (mappedPositions.length > 0) {
      const start = Math.min(...mappedPositions);
      const end = Math.max(...mappedPositions) + 1;
      const txt = scriptText.slice(start, end).trim();
      if (txt.length > 0) {
        corrected.push({ text: txt, start: seg.start, end: seg.end });
      }
    }
    wp += segLen;
  }

  console.log(`     ✅ Aligned: ${corrected.length} segments`);

  // Step 5: Build SRT (split long segments into ≤2 lines)
  const srtBlocks: string[] = [];
  let srtIdx = 1;

  for (const seg of corrected) {
    let text = seg.text;
    // Remove any remaining periods
    text = text.replace(/\./g, "").trim();
    if (!text) continue;

    // Split long text into max 2 lines of ~45 chars
    if (text.length > 90) {
      const mid = Math.floor(text.length / 2);
      const breakAt = text.indexOf(" ", mid);
      if (breakAt > 0 && breakAt < text.length - 5) {
        text = text.slice(0, breakAt) + "\n" + text.slice(breakAt + 1);
      }
    } else if (text.length > 45) {
      const mid = Math.floor(text.length / 2);
      const breakAt = text.indexOf(" ", mid);
      if (breakAt > 0) {
        text = text.slice(0, breakAt) + "\n" + text.slice(breakAt + 1);
      }
    }

    srtBlocks.push(`${srtIdx}\n${formatSrtTime(seg.start)} --> ${formatSrtTime(seg.end)}\n${text}`);
    srtIdx++;
  }

  const srt = srtBlocks.join("\n\n") + "\n";

  // Save
  const contentDir = join(ROOT, "content", date, mosque);
  const srtPath = join(contentDir, "subtitle-th.srt");
  writeFileSync(srtPath, srt);
  console.log(`  ✅ Generated ${srtIdx - 1} subtitle blocks → ${srtPath}`);

  return srtPath;
}

// CLI
if (import.meta.main) {
  const date = process.argv[2];
  const mosque = process.argv[3];
  if (!date || !mosque) {
    console.log("Usage: bun pipeline/subtitle-gen.ts <date> <mosque>");
    process.exit(1);
  }
  generateSubtitle(date, mosque);
}

File: pipeline/subtitle-gen.py
#!/usr/bin/env python3
"""
Generate Thai SRT subtitle from Whisper timing + translation text

Uses SequenceMatcher to align Whisper transcription with our translation,
then splits at Thai word boundaries using pythainlp.

Usage: python3 pipeline/subtitle-gen.py <date> <mosque>
"""

import json, re, sys, subprocess
from pathlib import Path
from difflib import SequenceMatcher
from pythainlp.tokenize import word_tokenize as th_tokenize

ROOT = Path(__file__).parent.parent
AUDIO_DIR = ROOT / "output" / "audio"


def fmt_srt(seconds):
    h = int(seconds // 3600)
    m = int((seconds % 3600) // 60)
    s = int(seconds % 60)
    ms = int((seconds % 1) * 1000)
    return f"{h:02d}:{m:02d}:{s:02d},{ms:03d}"


def word_boundaries(text):
    tokens = th_tokenize(text, engine="newmm")
    boundaries = set()
    p = 0
    for tok in tokens:
        boundaries.add(p)
        p += len(tok)
    boundaries.add(p)
    return sorted(boundaries)


def nearest_boundary(text, pos):
    bounds = word_boundaries(text)
    return min(bounds, key=lambda b: abs(b - pos))


def get_clean_text(date, mosque):
    # Use TTS source text if available (exact text that TTS read)
    tts_source = ROOT / "content" / date / mosque / "tts-source-text.txt"
    if tts_source.exists():
        return tts_source.read_text().strip()

    path = ROOT / "content" / date / mosque / "translation-th.md"
    text = path.read_text()

    # Remove frontmatter
    if text.startswith("---\n"):
        end = text.find("\n---\n", 4)
        if end > 0:
            text = text[end + 5:]

    # Strip markdown
    text = re.sub(r'^#{1,6}\s+.*$', '', text, flags=re.MULTILINE)
    text = re.sub(r'^---+$', '', text, flags=re.MULTILINE)
    text = re.sub(r'.*ส่วนที่\s+\d+/\d+.*', '', text)
    text = re.sub(r'^-?\s*\*{0,2}(วันที่|เชค|แหล่งที่มา|Date|Sheikh|Source).*$', '', text, flags=re.MULTILINE | re.IGNORECASE)
    text = re.sub(r'https?://\S+', '', text)
    text = re.sub(r'\[([^\]]*)\]\([^)]*\)', r'\1', text)
    text = re.sub(r'## ไฮไลท์สำหรับ Reels[\s\S]*$', '', text)
    text = re.sub(r'\*\*(.+?)\*\*', r'\1', text)
    text = re.sub(r'\*(.+?)\*', r'\1', text)
    text = re.sub(r'^>\s*', '', text, flags=re.MULTILINE)
    text = re.sub(r'^[-*]\s+', '', text, flags=re.MULTILINE)
    # Remove periods
    text = re.sub(r'\.(\n)', r'\1', text)
    text = re.sub(r'\. ', ' ', text)
    text = re.sub(r'\.$', '', text, flags=re.MULTILINE)
    # Flatten
    text = re.sub(r'\s+', ' ', text).strip()
    return text


def run_whisper(audio_path, output_dir):
    wav_path = output_dir / "whisper_input.wav"
    json_path = output_dir / "whisper_input.json"

    if not wav_path.exists():
        subprocess.run([
            "ffmpeg", "-y", "-i", str(audio_path),
            "-ar", "16000", "-ac", "1", str(wav_path)
        ], capture_output=True, check=True)

    if not json_path.exists():
        print("     🎤 Running mlx_whisper turbo...")
        subprocess.run([
            "mlx_whisper", str(wav_path),
            "--model", "mlx-community/whisper-turbo",
            "--language", "th",
            "-f", "json",
            "-o", str(output_dir),
            "--condition-on-previous-text", "False",
            "--hallucination-silence-threshold", "2",
            "--no-speech-threshold", "0.5",
        ], check=True)

    return json.loads(json_path.read_text())


def main():
    if len(sys.argv) < 3:
        print("Usage: python3 pipeline/subtitle-gen.py <date> <mosque>")
        sys.exit(1)

    date, mosque = sys.argv[1], sys.argv[2]
    tts_audio = AUDIO_DIR / f"{date}-{mosque}-th.mp3"

    if not tts_audio.exists():
        print(f"  ❌ TTS audio not found: {tts_audio}")
        sys.exit(1)

    # 1. Whisper
    whisper_dir = AUDIO_DIR / f"{date}-{mosque}-whisper"
    whisper_dir.mkdir(parents=True, exist_ok=True)
    result = run_whisper(tts_audio, whisper_dir)
    segments = result["segments"]
    print(f"     📊 Whisper: {len(segments)} segments")

    # 2. Get script text
    script_text = get_clean_text(date, mosque)
    print(f"     📝 Script: {len(script_text)} chars")

    # 2b. Remove Whisper hallucinations (30s garbage segments)
    clean_segments = []
    for seg in segments:
        t = seg["text"].strip()
        d = seg["end"] - seg["start"]
        words = t.split()
        unique_words = len(set(words))
        if d >= 20 and unique_words <= 6 and len(words) > 8:
            print(f"     ⚠️ Hallucination removed: [{seg['start']:.0f}s-{seg['end']:.0f}s]")
            continue
        clean_segments.append(seg)
    segments = clean_segments

    # 3. Parse segments — split long ones at ~3.5s boundaries
    parsed = []
    for seg in segments:
        t = seg["text"].strip()
        d = seg["end"] - seg["start"]
        if d <= 5 or len(t) <= 40:
            parsed.append({"text": t, "start": seg["start"], "end": seg["end"]})
        else:
            # Split proportionally
            n_parts = max(2, int(d / 3.5))
            part_len = len(t) / n_parts
            part_dur = d / n_parts
            for i in range(n_parts):
                s_idx = int(i * part_len)
                e_idx = int((i + 1) * part_len) if i < n_parts - 1 else len(t)
                # Snap to Thai word boundary
                if i > 0:
                    s_idx = nearest_boundary(t, s_idx)
                if i < n_parts - 1:
                    e_idx = nearest_boundary(t, e_idx)
                chunk = t[s_idx:e_idx].strip()
                if chunk:
                    parsed.append({
                        "text": chunk,
                        "start": seg["start"] + i * part_dur,
                        "end": seg["start"] + (i + 1) * part_dur
                    })

    print(f"     📐 Parsed: {len(parsed)} segments")

    # 4. SequenceMatcher alignment
    print("     🔗 Aligning whisper ↔ script...")
    wf = "".join(s["text"] for s in parsed)
    m = SequenceMatcher(None, wf, script_text, autojunk=False)
    w2s = {}
    for tag, w1, w2, s1, s2 in m.get_opcodes():
        if tag == "equal":
            for i, j in zip(range(w1, w2), range(s1, s2)):
                w2s[i] = j
        elif tag == "replace":
            for k in range(w2 - w1):
                w2s[w1 + k] = s1 + int(k * (s2 - s1) / max(w2 - w1, 1))
        elif tag == "delete":
            for k in range(w1, w2):
                w2s[k] = s1

    # 5. Replace whisper text with script text
    corrected = []
    wp = 0
    for seg in parsed:
        sl = len(seg["text"])
        sp = [w2s[i] for i in range(wp, wp + sl) if i in w2s]
        if sp:
            start_in_script = min(sp)
            end_in_script = max(sp) + 1
            start_snapped = nearest_boundary(script_text, start_in_script)
            end_snapped = nearest_boundary(script_text, end_in_script)
            txt = script_text[start_snapped:end_snapped].strip()
            if txt:
                corrected.append({"text": txt, "start": seg["start"], "end": seg["end"]})
        wp += sl

        # Sort by time
        corrected.sort(key=lambda x: x["start"])

    print(f"     ✅ Corrected: {len(corrected)} segments")

    # 5b. Post-process: split segments that are too long (>80 chars or >6 seconds)
    final_segments = []
    for seg in corrected:
        text = seg["text"].replace(".", "").strip()
        dur = seg["end"] - seg["start"]
        if not text:
            continue
        if len(text) <= 80 and dur <= 6:
            final_segments.append({"text": text, "start": seg["start"], "end": seg["end"]})
        else:
            # Split at Thai word boundaries
            bounds = word_boundaries(text)
            n_parts = max(2, len(text) // 50)
            part_dur = dur / n_parts
            part_len = len(text) / n_parts
            for i in range(n_parts):
                s_idx = nearest_boundary(text, int(i * part_len)) if i > 0 else 0
                e_idx = nearest_boundary(text, int((i + 1) * part_len)) if i < n_parts - 1 else len(text)
                chunk = text[s_idx:e_idx].strip()
                if chunk:
                    final_segments.append({
                        "text": chunk,
                        "start": seg["start"] + i * part_dur,
                        "end": seg["start"] + (i + 1) * part_dur
                    })

    print(f"     📐 Final: {len(final_segments)} segments (after split)")

    # 6. Build SRT
    srt_blocks = []
    for i, seg in enumerate(final_segments, 1):
        text = seg["text"].strip()
        if not text:
            continue

        # Split into max 2 lines at word boundary
        if len(text) > 45:
            mid = len(text) // 2
            break_at = nearest_boundary(text, mid)
            if 10 < break_at < len(text) - 10:
                text = text[:break_at].strip() + "\n" + text[break_at:].strip()

        srt_blocks.append(f"{i}\n{fmt_srt(seg['start'])} --> {fmt_srt(seg['end'])}\n{text}")

    srt = "\n\n".join(srt_blocks) + "\n"

    # Save
    srt_path = ROOT / "content" / date / mosque / "subtitle-th.srt"
    srt_path.write_text(srt)
    print(f"  ✅ Generated {len(srt_blocks)} subtitle blocks → {srt_path}")


if __name__ == "__main__":
    main()
